File: cli/python/src/memgo_cli/cli/commands/memory/delete.py
"""记忆 delete 用例。"""

from __future__ import annotations

import time as _time

import typer
from rich.console import Console

from memgo_cli.backend.types import Backend
from memgo_cli.output.branding import (
    print_error,
    print_info,
    print_success,
    timed_status,
)
from memgo_cli.output.format import (
    format_agent_envelope,
    format_json,
    format_single_memory,
)

console = Console()
err_console = Console(stderr=True)


def cmd_delete(
    backend: Backend,
    memory_id: str,
    *,
    dry_run: bool,
    force: bool,
    delete_linked: bool,
    output: str,
) -> None:
    """按 ID 删除单条记忆，支持预览和确认。"""
    from memgo_cli.runtime.state import is_agent_mode, set_current_command

    set_current_command("delete")
    if is_agent_mode():
        output = "agent"
    if dry_run:
        # 读取并显示预期删除的记忆
        try:
            mem = backend.get(memory_id)
        except Exception as e:
            print_error(err_console, str(e), hint=None)
            raise typer.Exit(1) from None
        format_single_memory(console, mem, output)
        print_info(console, "No changes made (dry run).")
        return

    _start = _time.perf_counter()
    with timed_status(err_console, "Deleting...") as _ts:
        try:
            result = backend.delete(
                memory_id=memory_id,
                delete_linked=delete_linked,
                all=False,
                user_id=None,
                agent_id=None,
                app_id=None,
                run_id=None,
            )
        except Exception as e:
            print_error(err_console, str(e), hint=None)
            raise typer.Exit(1) from None
    _elapsed = _time.perf_counter() - _start

    if output == "agent":
        format_agent_envelope(
            console,
            command="delete",
            data={"id": memory_id, "deleted": True},
            duration_ms=int(_elapsed * 1000),
            scope=None,
            count=None,
        )
    elif output == "json":
        format_json(console, result)
    elif output != "quiet":
        print_success(console, f"Memory {memory_id[:8]} deleted ({_elapsed:.2f}s)")

File: cli/node/src/cli/commands/memory/delete.ts
import type { JsonObject } from '../../../backend/json.js';
import type { Backend } from '../../../backend/types.js';
import { printError, printInfo, printSuccess, timedStatus } from '../../../output/branding.js';
import { formatAgentEnvelope, formatJson } from '../../../output/format.js';
import { exit } from '../../../runtime/exit.js';
import { setCurrentCommand } from '../../../runtime/state.js';

/**
 * 确认后删除单条记忆，预演时仅显示目标。
 * @param backend - 已装配的后端连接器。
 * @param memoryId - 目标记忆 ID。
 * @param opts - 本次操作选项。
 * @param opts.output - 终端输出格式。
 * @param opts.dryRun - 仅展示删除目标，不执行删除。
 * @param opts.force - 跳过已有配置或删除操作的交互确认。
 * @param opts.deleteLinked - 同时删除关联记忆的选项。
 */
export async function cmdDelete(
  backend: Backend,
  memoryId: string,
  opts: {
    output: string;
    dryRun?: boolean;
    force?: boolean;
    deleteLinked?: boolean;
  }
): Promise<void> {
  setCurrentCommand('delete');
  if (opts.dryRun) {
    let mem: JsonObject;
    try {
      mem = await backend.get(memoryId);
    } catch (e) {
      printError(e instanceof Error ? e.message : String(e));
      exit(1);
    }
    const text = (mem.memory ?? mem.text ?? '') as string;
    printInfo(`Would delete memory ${memoryId.slice(0, 8)}: ${text}`);
    printInfo('No changes made.');
    return;
  }

  const start = performance.now();
  let result: JsonObject;
  try {
    result = await timedStatus('Deleting...', async () => {
      return backend.delete(memoryId, { deleteLinked: opts.deleteLinked });
    });
  } catch (e) {
    printError(e instanceof Error ? e.message : String(e));
    exit(1);
  }
  const elapsed = (performance.now() - start) / 1000;

  if (opts.output === 'agent') {
    formatAgentEnvelope({
      command: 'delete',
      data: { id: memoryId, deleted: true },
      durationMs: Math.round(elapsed * 1000),
    });
  } else if (opts.output === 'json') {
    formatJson(result);
  } else if (opts.output !== 'quiet') {
    printSuccess(`Memory ${memoryId.slice(0, 8)} deleted (${elapsed.toFixed(2)}s)`);
  }
}
